src/validation.cpp

```diff
@@ -3406,6 +3405,33 @@ static CDiskBlockPos SaveBlockToDisk(const CBlock& block, int nHeight, const CCh
     return blockPos;
 }
 
+bool AcceptStake(CBlockIndex &index, const CBlock &block, CValidationState &state, const CCoinsViewCache &coins_cache) {
+    const auto stake_validator = GetComponent<staking::StakeValidator>();
+    UTXOViewAdapter utxo_view(GetComponent<staking::ActiveChain>(), coins_cache);
+
+    staking::BlockValidationResult stake_validation_result =
+        stake_validator->CheckStake(block, &state.block_validation_info, CheckStakeFlags::NONE, &utxo_view);
+
+    if (!staking::CheckResult(stake_validation_result, state)) {
+        LogPrint(BCLog::VALIDATION, "%s: Invalid stake found for block=%s failure=%s\n",
+                 __func__, block.GetHash().ToString(), stake_validation_result.GetRejectionMessage());
+        return false;
+    }
+
+    const CTransactionRef coinbase = block.vtx[0];
+    PrecomputedTransactionData txdata(*coinbase);
+
+    constexpr unsigned int standard_flags = STANDARD_SCRIPT_VERIFY_FLAGS;
+    if (!CheckInputs(*coinbase, state, coins_cache, true, standard_flags, true, true, txdata)) {
+        LogPrint(BCLog::VALIDATION, "%s: Invalid stake found for block=%s failure=%s\n",
```

Comment:
Nit: it's better to say `invalid coinbase input`, coinbase transaction can also have non-stake inputs.